src/baldr/graphtile.cc

```diff
@@ -69,29 +66,54 @@ GraphTile::GraphTile(const std::string& tile_dir, const GraphId& graphid) : head
     size_t filesize = file.tellg();
     graphtile_.reset(new std::vector<char>(filesize));
     file.seekg(0, std::ios::beg);
-    file.read(&(*graphtile_)[0], filesize);
+    file.read(graphtile_->data(), filesize);
     file.close();
 
     // Set pointers to internal data structures
-    Initialize(graphid, &(*graphtile_)[0], graphtile_->size());
-  } else {
-    std::ifstream file(file_location + ".gz", std::ios::in | std::ios::binary | std::ios::ate);
+    Initialize(graphid, graphtile_->data(), graphtile_->size());
+  } // try to load a gzipped tile
+  else {
+    std::ifstream file(file_location + ".gz", std::ios::in | std::ios::binary);
     if (file.is_open()) {
-      // Pre-allocate assuming 3.25:1 compression ratio (based on scanning some large NA tiles)
-      size_t filesize = file.tellg();
-      file.seekg(0, std::ios::beg);
-      graphtile_.reset(new std::vector<char>());
-      graphtile_->reserve(filesize * 3 +
-                          filesize / 4); // TODO: read the gzip footer and get the real size?
+      // read the compressed file into memory
+      std::vector<char> compressed(std::istreambuf_iterator<char>(file),
```

Comment:
both inflating and deflating work by providing src and dst functions to the method that does the work. the src function gets more src data into a buffer for zlib to read from where the dst function puts zlibs output to its final resting place. there is one tricky bit with the dst function which is you dont always need more buffer for the next round if there is no more src data to be processed (hence the `if` below)